File: Leetcode/0109 Convert Sorted List to Binary Search Tree.cpp
class Solution {
public:
    // O(n log n) time, O(log n) space
    // does not modify the list 
    TreeNode* sortedListToBST(ListNode* head, ListNode* end=nullptr) {
        if(head == end) {
            return nullptr;
        }
        if(head->next == end) {
            return new TreeNode(head->val);
        }
        ListNode *slow = head, *fast = head;
        while (fast != end and fast->next != end) {
            slow = slow->next;
            fast = fast->next->next;
        }
        TreeNode *root = new TreeNode(slow->val);
        root->left = sortedListToBST(head, slow);
        root->right = sortedListToBST(slow->next, end);
        return root;
    }
};

File: Leetcode/0109 Convert Sorted List to Binary Search Tree.py
class Solution:
    # does not modify the lsit
    # O(n log n) time, O(log n) stack space
    def sortedListToBST(self, head, end=None):
        if head is end:
            return None

        if head.next is end:
            return TreeNode(head.val)

        slow, fast = head, head
        while (fast is not end) and (fast.next is not end):
            slow = slow.next
            fast = fast.next.next

        root = TreeNode(slow.val)
        root.left = self.sortedListToBST(head, slow)
        root.right = self.sortedListToBST(slow.next, end)
        return root